[Misc] PR templates (#3413)
vllm-project/vllm: .github/PULL_REQUEST_TEMPLATE.md

## PR Checklist

Before submitting your pull request, please ensure you have met the following criteria. This helps us maintain quality and efficiency in our review process. Check each box to indicate your understanding and compliance:

### PR Title And Classification

- [ ] **I understand that** only specific types of PRs will be reviewed. My PR title is prefixed appropriately to indicate the type of change, using one of the following:

  - [ ] `[Doc]` for documentation fixes.
  - [ ] `[Bugfix]` for bug fixes.
  - [ ] `[CI/Build]` for build or continuous integration improvements.
  - [ ] `[Model]` for adding a new model or improving an existing model. Model name should appear in the title.
  - [ ] `[Kernel]` for changes affecting computation kernels.
  - [ ] `[Hardware][Vendor]` for hardware-specific changes. Vendor name should appear in the prefix, e.g., `[Hardware][AMD]`.
  - [ ] `[Misc]` for PRs that do not fit the above categories. Please use this sparingly.

  - **Note:** If my PR spans more than one category, I will include all relevant prefixes.

### Coding Standards

> - [ ] **I understand that** I must run `./format.sh` **before submitting the PR and after any new commits** to ensure compliance with linter checks. PRs failing to meet linter standards will not be reviewed or merged.

Reviewer: ```suggestion
- [ ] **I understand that** I must run `./format.sh` **before submitting the PR and after any new commits** to ensure compliance with linter checks. PRs failing to meet linter standards will not be merged.
```

Author: Fixed in [289fa60](https://github.com/vllm-project/vllm/pull/3413/commits/289fa6057896be9d868ddb71082c08f056af2e29).

<details>
<!-- inside this <details> section, markdown rendering does not work, so we have to use raw html here. -->

> <summary><h2> PR Checklist (Click to expand. Please do have a look before submitting a PR!) </h2></summary>

Reviewer 2: ```suggestion
<summary><h2> PR Checklist (Click to expand. Please take a look before submitting a PR!) </h2></summary>
```

<p>Before submitting your pull request, please ensure you have met the following criteria. This helps us maintain quality and efficiency in our review process.</p>

<h3>PR Title And Classification</h3>
<p><strong>I understand that</strong> only specific types of PRs will be reviewed. My PR title is prefixed appropriately to indicate the type of change, using one of the following:</p>
<ul>
    <li><code>[Doc]</code> for documentation fixes.</li>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>

>     <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>

Reviewer 2: ```suggestion
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core vLLM (e.g., <code>LLMEngine</code>, <code>Scheduler</code>, etc.)</li>
```

    <li><code>[Kernel]</code> for changes affecting computation kernels.</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix, e.g., <code>[Hardware][AMD]</code>.</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If my PR spans more than one category, I will include all relevant prefixes.</p>

<h3>Coding Standards</h3>

> <p><strong>I understand that</strong> I must run <code>./format.sh</code> <strong>before submitting the PR and after any new commits</strong> to ensure compliance with linter checks. PRs failing to meet linter standards will not be merged.</p>

Reviewer 2: ```suggestion
<p><strong>I understand that</strong> I should run <code>format.sh</code> <strong>before submitting the PR and after any new commits</strong> to ensure compliance with linter checks. PRs failing to meet linter standards will not be merged.</p>
```

<h3>Code Quality</h3>

> <p><strong>I understand that</strong> my code must be well-commented and include sufficient tests, ensuring future contributors can easily understand and modify the codebase.</p>

Reviewer 2: Can you add a check-box list here like the following:

- [ ] The PR is well-documented in the code, as well as in the document (`docs/source/`) if it adds new features.
- [ ] The PR adds unit tests for the new components.
- [ ] The PR adds integration tests for the new features.

Reviewer 2: Also this check-box list should not be hided in the details.

Author: I use <details> here because @simon-mo wants PRs to look clean.

I think that is also reasonable, as these checklists are meant for pr submitters to read and check, they are not very meaningful to others.

Author: Overall we should trust the PR submitters. These checklists are more of **reminders** for them, rather than some items to be checked by the vllm team.

For a checkbox-like PR template, you can take a look at https://github.com/open-mmlab/mmcv/pull/3057 . Indeed it makes the PR filled with unmeaningful information. Collapsing them by default is better in my opinion.

<h3>Documentation for User-Facing Changes</h3>
<p><strong>I understand that</strong> if my PR introduces user-facing changes, it must be accompanied by relevant documentation to help users understand and utilize the new features or changes.</p>

<p>Thank you for your contribution!</p>

</details>

## PR Justification

Please provide a brief explanation of the motivation behind the PR and the changes it introduces. This helps reviewers understand the context and rationale for the contribution.

If possible, please link existing issues this PR will resolve.

## Note for large changes

> For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue discussing the technical design and justification. Otherwise, we will tag it with `rfc-required` and might not take a look at the PR.

Reviewer 2: ```suggestion
For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue discussing the technical design and justification. Otherwise, we will tag it with `rfc-required` and might not go through the PR.
```

    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>

>     <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>

Author: @zhuohan123 I think this linter check thing has to be highlighted. I'm thinking of using red color and bold text.

Reviewer 2: I personally feel this is fine? CI also tests this so I'm not too worried

Author: Okay, then you can merge it. LGTM.
